app/components/autocomplete/at_mention/index.js

```diff
@@ -29,16 +51,34 @@ function mapStateToProps(state, ownProps) {
         }
     }
 
+    const match = postDraft.substring(0, cursorPosition).match(regex);
+    let matchTerm = null;
+    if (match) {
+        matchTerm = isSearch ? match[1] : match[2];
+    }
+
+    const opts = {
+        matchTerm,
+        currentUserId,
+        isSearch
+    };
+
+    if (isSearch) {
+        filterMembersInCurrentTeam(state, {...opts, array: teamMembers});
+    } else {
+        filterMembersInChannel(state, {...opts, array: inChannel});
+        filterMembersNotInChannel(state, {...opts, array: outChannel});
+    }
+
     return {
-        currentUserId: state.entities.users.currentUserId,
-        currentChannelId,
+        currentUserId,
+        currentChannelId: isSearch ? '' : currentChannelId,
         currentTeamId: state.entities.teams.currentTeamId,
         defaultChannel: getDefaultChannel(state),
+        hasMatch: match !== null,
+        matchTerm,
         postDraft,
-        autocompleteUsers: {
-            inChannel: getProfilesInCurrentChannel(state),
-            outChannel: getProfilesNotInCurrentChannel(state)
-        },
+        autocompleteUsers: autocompleteProfiles,
```

Comment:
This isn't good (although the old way also had issues). We're mutating the value passed to autocompleteUsers so AtMention won't know when to re-render